File: benchmarks/python/03_recursion.py
import sqlite3
import time
import os
import sys
from bench_utils import get_connection, load_plsqlite

def bench_app_layer(depth=100):
    conn = get_connection()
    conn.execute("DROP TABLE IF EXISTS recursion_log")
    conn.execute("CREATE TABLE recursion_log(depth INTEGER, val INTEGER)")
    start = time.time()
    with conn:
        def recurse(n):
            if n <= 0: return 0
            conn.execute("INSERT INTO recursion_log(depth, val) VALUES (?, ?)", (n, n * 2))
            return n + recurse(n - 1)
        recurse(depth)
    end = time.time()
    res = conn.execute("SELECT count(*) FROM recursion_log").fetchone()[0]
    conn.close()
    return (end - start) * 1000, res

def bench_plsqlite(depth=100):
    conn = get_connection()
    load_plsqlite(conn)
    conn.execute("DROP TABLE IF EXISTS recursion_log")
    conn.execute("CREATE TABLE recursion_log(depth INTEGER, val INTEGER)")

    conn.execute("""
    SELECT register_plsql('rec_log_insert', 'n', '
        IF (@n <= 0) THEN
            RETURN 0;
        END IF;
        INSERT INTO recursion_log(depth, val) VALUES (@n, @n * 2);
        RETURN @n + (SELECT run_plsql("rec_log_insert", @n - 1));
    ');
    """)
    start = time.time()
    conn.execute("SELECT run_plsql('rec_log_insert', ?)", (depth,)).fetchone()
    end = time.time()
    res = conn.execute("SELECT count(*) FROM recursion_log").fetchone()[0]
    conn.close()
    return (end - start) * 1000, res

if __name__ == "__main__":
    print("--- Python Recursion Benchmarks ---")
    depth = 100
    app_time, app_res = bench_app_layer(depth)
    print(f"RESULT: Recursion: App-Layer: {app_time:.4f}ms (Result: {app_res})")

    pl_time, pl_res = bench_plsqlite(depth)
    print(f"RESULT: Recursion: PL/SQLite Proc: {pl_time:.2f}ms (Result: {pl_res})")

    if pl_time > 0:
        print(f"Speedup: {app_time/pl_time:.4f}x")

File: benchmarks/node/03_recursion.js
const { getDbConnection, loadPLSQLite } = require('./bench_utils');

function benchAppLayer(depth = 100) {
    const db = getDbConnection();
    db.prepare("DROP TABLE IF EXISTS recursion_log").run();
    db.prepare("CREATE TABLE recursion_log(depth INTEGER, val INTEGER)").run();
    const start = process.hrtime.bigint();

    const insertStmt = db.prepare("INSERT INTO recursion_log(depth, val) VALUES (?, ?)");
    function recurse(n) {
        if (n <= 0) return 0;
        insertStmt.run(n, n * 2);
        return n + recurse(n - 1);
    }

    db.transaction(() => recurse(depth))();
    const end = process.hrtime.bigint();
    const count = db.prepare("SELECT count(*) as c FROM recursion_log").get().c;
    db.close();
    return { time: Number(end - start) / 1000000, res: count };
}

function benchPLSQLite(depth = 100) {
    const db = getDbConnection();
    loadPLSQLite(db);
    db.prepare("DROP TABLE IF EXISTS recursion_log").run();
    db.prepare("CREATE TABLE recursion_log(depth INTEGER, val INTEGER)").run();

    db.prepare(`
        SELECT register_plsql('rec_log_insert', 'n', '
            IF (@n <= 0) THEN 
                RETURN 0; 
            END IF;
            INSERT INTO recursion_log(depth, val) VALUES (@n, @n * 2);
            RETURN @n + (SELECT run_plsql(''rec_log_insert'', @n - 1));
        ');
    `).run();

    const start = process.hrtime.bigint();
    db.prepare("SELECT run_plsql('rec_log_insert', ?)").run(depth);
    const end = process.hrtime.bigint();
    const count = db.prepare("SELECT count(*) as c FROM recursion_log").get().c;
    db.close();
    return { time: Number(end - start) / 1000000, res: count };
}

console.log("--- Node.js Recursion Benchmarks ---");
const depth = 100;

const app = benchAppLayer(depth);
console.log(`RESULT: Recursion: App-Layer: ${app.time.toFixed(4)}ms (Result: ${app.res})`);

const pl = benchPLSQLite(depth);
console.log(`RESULT: Recursion: PL/SQLite Proc: ${pl.time.toFixed(2)}ms (Result: ${pl.res})`);

if (pl.time > 0) {
    console.log(`Speedup: ${(app.time / pl.time).toFixed(4)}x`);
}
